feat: enable random anti-fee sniping

aagbotemi opened the pull request:
This PR implements Anti-Fee-Sniping with randomization as discussed in issue #4 

## Notes to the reviewers
The implementation adds randomization to anti-fee-sniping behavior:

1. Uses a 50/50 chance to choose between nLockTime and nSequence (when possible)
2. Adds a 10% chance to set either value further back in time (by a random value between 0-99)
3. Detects taproot inputs and their confirmation status



## Changelog notice

### Changed
- `CreatePsbtError::MissingFullTxForLegacyInput` and `CreatePsbtError::MissingFullTxForSegwitV0Input` now wrap `Input` in `Box` (breaking change)

### Added
- `PsbtParams::enable_anti_fee_sniping` field for BIP326 anti-fee-sniping protection
- `CreatePsbtError::InvalidLockTime` and `CreatePsbtError::UnsupportedVersion` error variants
- `Selection::create_psbt_with_rng` method for custom RNG



* [x] I've signed all my commits
* [x] I ran `cargo fmt` and `cargo clippy` before committing
* [x] I've added docs for the new feature

Closes #4

ValuedMammal commented:
Thanks @aagbotemi I'm still in the process of reviewing. I agree it would be nice to resolve the clippy error - my other suggestion would be to globally allow the lints in `lib.rs` for now and open an issue to fix them in a future PR, but in general I agree with your approach to simply `Box` the large enum variants.

aagbotemi commented:
> Thanks @aagbotemi I'm still in the process of reviewing.

Thank you @ValuedMammal. Clippy error and large enum variant error in the CI build has been fixed. For subsequent ones(if there is), I'll open another for it.

ValuedMammal commented on src/utils.rs:
```diff
@@ -0,0 +1,123 @@
+                .max(MIN_SEQUENCE_VALUE);
+        }
+
+        tx.input[*idx].sequence = Sequence(sequence_value);
```
To be clear, this seems to be making an assumption that the index found by enumerating the selection inputs corresponds to the same TxIn in `tx.input`, right?

aagbotemi replied:
Yes, given that the length of `tx.input` is the same as `inputs`. I have now made a measure for correctness and safety - ensuring that there is right data with the right indices.

ValuedMammal replied:
Ok good point, the length must be the same or else we would be in the "use LockTime" branch.

Still, to not have the implicit ordering assumption I thought it would be better to find the correct input by outpoint, by maybe I'm over-analyzing.

<details>
  <summary>diff</summary>

```diff
-    let taproot_inputs: Vec<usize> = (0..tx.input.len())
-        .filter(|&idx| {
-            // Check if this input is taproot using the corresponding Input data
-            inputs
-                .get(idx)
-                .and_then(|input| input.plan())
-                .and_then(|plan| plan.witness_version())
-                .map(|version| version == WitnessVersion::V1)
-                .unwrap_or(false)
+    // vector of input_index and associated Input ref.
+    let taproot_inputs: Vec<(usize, &Input)> = tx
+        .input
+        .iter()
+        .enumerate()
+        .filter_map(|(vin, txin)| {
+            let input = inputs
+                .iter()
+                .find(|input| input.prev_outpoint() == txin.previous_output)?;
+            if input.prev_txout().script_pubkey.is_p2tr() {
+                Some((vin, input))
+            } else {
+                None
+            }
         })
         .collect();
 
     // Check always‐locktime conditions
@@ -70,10 +75,11 @@ pub fn apply_anti_fee_sniping(
         tx.lock_time = new_locktime;
     } else {
         // Use Sequence
         tx.lock_time = LockTime::ZERO;
-        let input_index = random_range(&mut rng, taproot_inputs.len() as u32) as usize;
-        let confirmation = inputs[input_index].confirmations(current_height);
+        let random_index = random_range(&mut rng, taproot_inputs.len() as u32);
+        let (input_index, input) = taproot_inputs[random_index as usize];
+        let confirmation = input.confirmations(current_height);
```
</details>

aagbotemi replied:
You are absolutely correct. Your code is robust and helps the anti-fee-sniping logic to operate on the right inputs. 

Thank you for being thorough and helping the code to be correct and more maintainable.

ValuedMammal reviewed (commented):
I have some review notes

- For imports, prefer `alloc` over `std`. This way we can actually support no-std.
- I want to see a test of some kind that creates a PSBT with `enable_anti_fee_sniping` and checks the expected values of the locktime and/or sequence.

aagbotemi commented on Cargo.toml:
```diff
@@ -11,14 +11,15 @@ readme = "README.md"
 [dependencies]
 miniscript = { version = "12", default-features = false }
 bdk_coin_select = "0.4.0"
+rand_core = { version = "0.6.0", features = ["getrandom"] }
+bdk_chain = { version = "0.21" }
+bitcoin = { version = "0.32", features = ["rand-std"] }
```
`bdk_chain` was made an explicit dependency, because I used it in the first commit before pr #1 was merged, and i had to refactor the code. I have added it to `dev-dependencies`

`bitcoin` dep has been made to have no default feature, and rust-bitcoin types have been imported via `miniscript`. though the types are still been imported from `bitcoin` in synopsis and common examples

aagbotemi commented on src/selection.rs:
```diff
@@ -53,6 +56,7 @@ impl Default for PsbtParams {
             fallback_locktime: absolute::LockTime::ZERO,
             fallback_sequence: FALLBACK_SEQUENCE,
             mandate_full_tx_for_segwit_v0: true,
+            enable_anti_fee_sniping: true,
```
okay, that's true

aagbotemi commented on src/selection.rs:
```diff
@@ -146,8 +160,23 @@ impl Selection {
+            let current_height = bitcoin::absolute::Height::from_consensus(height)
+                .map_err(|_conversion_error| CreatePsbtError::InvalidHeight(height))?;
+
+            apply_anti_fee_sniping(&mut tx, &self.inputs, current_height, rbf_enabled);
```
Yes. I have fixed that

aagbotemi commented on src/utils.rs:
```diff
@@ -0,0 +1,123 @@
+    (rand_val / max_u32) < probability
+}
+
+fn random_range(min: u32, max: u32) -> u32 {
```
Yes, exactly, I was trying to avoid modulo bias, though it's very tiny. I will update the code with the simpler version.

ValuedMammal commented on Cargo.toml:
```diff
@@ -32,3 +33,6 @@ name = "synopsis"
 crate-type = ["lib"]
+
+[[example]]
+name = "anti_fee_snipping"
```
Can we rename this to `anti_fee_sniping`.

aagbotemi replied:
Alright.

ValuedMammal commented on src/selection.rs:
```diff
@@ -63,13 +70,19 @@ pub enum CreatePsbtError {
+    MissingFullTxForLegacyInput(Box<Input>),
     /// Missing tx for segwit v0 input.
-    MissingFullTxForSegwitV0Input(Input),
+    MissingFullTxForSegwitV0Input(Box<Input>),
```
q: Should we change to `MissingFullTxForSegwitV0Input(OutPoint)` instead of having the `Box<Input>`?

aagbotemi replied:
I agree it could be refactored.

Would you prefer I address this now or in a follow-up PR?

ValuedMammal commented on src/utils.rs:
```diff
@@ -0,0 +1,98 @@
+    current_height: absolute::Height,
+    rbf_enabled: bool,
+) -> Result<(), CreatePsbtError> {
+    const MAX_SEQUENCE_VALUE: u32 = 65_535;
```
nit:
```suggestion
    const MAX_RELATIVE_HEIGHT: u32 = 65_535;
```

aagbotemi replied:
Alright.

ValuedMammal commented on src/utils.rs:
```diff
@@ -0,0 +1,98 @@
+    const USE_NLOCKTIME_PROBABILITY: u32 = 2;
+    const MIN_SEQUENCE_VALUE: u32 = 1;
+    const FURTHER_BACK_PROBABILITY: u32 = 10;
+    const MAX_RANDOM_OFFSET: u32 = 99;
```
Should this be `100` since we're doing the modulo?

aagbotemi replied:
Yes, it should be `100`. I have fixed it. Thank you.

ValuedMammal commented on src/utils.rs:
```diff
@@ -0,0 +1,98 @@
+        let confirmation = input.confirmations(current_height);
+        confirmation == 0
+            || confirmation > MAX_SEQUENCE_VALUE
+            || !matches!(
+                input.plan().and_then(|plan| plan.witness_version()),
+                Some(WitnessVersion::V1)
+            )
```
I was thinking it might be simpler and more robust to test if the input is taproot by using `Script::is_p2tr`, rationale being that we always have a prev txout but may not have a plan.

```suggestion
            || !input.prev_txout().script_pubkey.is_p2tr()
```

aagbotemi replied:
good, this approach is much better.

ValuedMammal commented on src/utils.rs:
```diff
@@ -0,0 +1,98 @@
+            locktime = locktime.saturating_sub(random_offset);
+        }
+
+        let new_locktime = LockTime::from_height(locktime)
+            .map_err(|_| CreatePsbtError::InvalidHeight(locktime))?;
```
I think we can expect it to be a valid LockTime because `current_height` has the type `absolute::Height`. And probably just remove the `CreatePsbtError::InvalidHeight` variant.

```suggestion
        let new_locktime = LockTime::from_height(locktime).expect("must be valid Height");
```

aagbotemi replied:
Alright.

ValuedMammal commented:
We should also be able to remove the clippy allow attributes in `lib.rs`.

aagbotemi commented:
> I'm still in favor of using `Box` on `PlanOrPsbtInput` and `ScriptSource::Descriptor`.

Fixed! I've updated the code to use `Box`.

aagbotemi commented on src/selection.rs:
```diff
@@ -63,33 +71,43 @@ pub enum CreatePsbtError {
+    MissingFullTxForLegacyInput(OutPoint),
     /// Missing tx for segwit v0 input.
-    MissingFullTxForSegwitV0Input(Input),
+    MissingFullTxForSegwitV0Input(OutPoint),
```
@ValuedMammal I have reverted the code.

aagbotemi commented on src/utils.rs:
```diff
@@ -0,0 +1,100 @@
+
+fn random_range(rng: &mut OsRng, max: u32) -> u32 {
+    rng.next_u32() % max
+}
```
Yes, this is good. It helps to solve modulos bias

I would like to know your thought on using
```rs
use bitcoin::secp256k1::rand::Rng;

fn random_range(rng: &mut OsRng, max: u32) -> u32 {
    rng.gen_range(0..max)
}
```

aagbotemi commented:
> In general I think the git history would be cleaner if we didn't merge with merge commits but instead rebase the commits in this PR.

Alright, I've done the cleanup

ValuedMammal commented:
Approach ACK.

nymius commented on src/utils.rs:
```diff
@@ -0,0 +1,157 @@
+            return n % end;
+        }
+    }
+}
```
Here you are defining the largest contiguous unbiased region within the numeric range representable by the u32 type. From what I have discovered, `(-end) % end` is the preferred version in the latest literature, although the performance is almost the same. Probably the compiler optimizes both forms to the same instructions, but I would use `(-end) % (end)` to fingerprint the technique you're using here for future code readers.

nymius commented on src/utils.rs:
```diff
@@ -0,0 +1,157 @@
+    random_range(rng, probability) == 0
+}
+
+// Return a random value in the range [0, end].
```
You are trying to add a docstring here or just a comment? Use `///` for docstrings.

nymius reviewed (commented):
Approach ACK 4649e8dc76ac97e6228d1628892dc82fee5d9cc5.

I guess this will be squashed before merging, but I would reserve the `revert` type in commit messages for those actually reverting other commits, not manual commits undoing something.

I also checked the references in the BIP and bitcoin core seems to have implemented something similar back in the time.

aagbotemi commented:
> Approach ACK [4649e8d](https://github.com/bitcoindevkit/bdk-tx/commit/4649e8dc76ac97e6228d1628892dc82fee5d9cc5).
> 


Thank you for the review. I'm working on fixing them.

aagbotemi commented on examples/anti_fee_sniping.rs:
```diff
@@ -0,0 +1,137 @@
+    // - Any input requires locktime (non-taproot, unconfirmed, or >65535 confirmations), OR
+    // - There are no taproot inputs, OR
+    // - Random 50/50 coin flip chose locktime
+    // - With 10% probability: subtracts random 0-99 block offset from current height
```
Noted, will fix that. Thank you

nymius commented on src/lib.rs:
```diff
@@ -1,8 +1,4 @@
 //! `bdk_tx`
-
-// FIXME: try to remove clippy "allows"
-#![allow(clippy::large_enum_variant)]
```
But they were introduced in a previous different PR, and the implications are general to all the repository, that's why I recommend to move them to a previous separated commit.

aagbotemi commented on src/utils.rs:
```diff
@@ -0,0 +1,156 @@
+    Ok(())
+}
+
+/// Returns true with probability 1/n.
```
The anti fee sniping implementation started from [here](https://github.com/bitcoindevkit/bdk/pull/1912#discussion_r2023587434). So, I have been using `rand_core` for the implementation since then. I can change to `rand` if we agree to use that.

cc: @ValuedMammal, we have access to using `Rng` trait using `rand`

aagbotemi committed:
refactor: remove unnecessary clippy allow attributes

aagbotemi committed:
feat!: implement anti-fee-sniping protection

BREAKING CHANGE:
- `CreatePsbtError::MissingFullTxForLegacyInput` and `CreatePsbtError::MissingFullTxForSegwitV0Input` now wrap `Input` in `Box`

Added new error variants to handle anti-fee-sniping specific failures:
- `InvalidLockTime` returns an error when the transaction locktime is time-based instead of height-based
- `UnsupportedVersion` returns an error when the transaction version is less than 2

nymius commented on examples/anti_fee_sniping.rs:
```diff
@@ -0,0 +1,139 @@
+        if height_val >= min_expected && height_val <= max_expected {
+            println!("✓ Locktime is within expected range");
+        } else {
+            println!("⚠ Locktime is outside expected range");
```
Is this line reachable? I guess you want to show the feature is working, but if anti fee sniping works as intended, this is not going to ever happen. This isn't a test. Maybe showing the `nLockTime` or `nSequence` in a meaningful way is a better approach for the example.

nymius replied:
Is this line reachable? I guess you want to show the feature is working, but if anti fee sniping works as intended, this is not going to ever happen. This isn't a test. Maybe showing the `nLockTime` or `nSequence` in a meaningful way is a better approach for the example.

nymius commented on examples/anti_fee_sniping.rs:
```diff
@@ -0,0 +1,152 @@
+
+    println!("nLockTime approach used: {} times", locktime_count);
+    println!("nSequence approach used: {} times", sequence_count);
+    println!("Both approaches provide anti-fee-sniping protection:");
```
nit: this isn't providing new information, I recommend to remove it

ValuedMammal commented on Cargo.toml:
```diff
@@ -13,11 +13,12 @@ readme = "README.md"
 [dependencies]
 miniscript = { version = "12", default-features = false }
 bdk_coin_select = "0.4.0"
```
Ideally we want to support no-std, so my question is how are we providing a source of randomness for no-std environments.

ValuedMammal replied:
The approach we took for bdk_wallet is to depend on `rand_core` in order to have `RngCore` trait and essentially split the interface based on whether `std` feature is enabled. If `std`, then we have access to `rand` via `bitcoin::secp256k1::rand`. If no-std then the user can pass in something that implements `RngCore`.

```rust
/// Create PSBT.
#[cfg(feature = "std")]
pub fn create_psbt(&self, params: PsbtParams) -> Result<bitcoin::Psbt, CreatePsbtError> {
    self.create_psbt_with_rng(params, &mut rand::thread_rng())
}

/// Create PSBT with `rng`.
pub fn create_psbt_with_rng(&self, params: PsbtParams, rng: &mut impl RngCore) -> Result<bitcoin::Psbt, CreatePsbtError> { ... }

```

The other option is to feature-gate the `enable_anti_fee_sniping` option entirely but that seems a bit less useful.

nymius replied:
@aagbotemi could you adapt your code [here](https://github.com/bitcoindevkit/bdk-tx/commit/782c3e367fb772d9ea094609a6b03fa46346bf25#diff-23af5356f6001cd3993cd3c801fb7716ea02d1e504081b9fb569332db6107e80R142-R156) and the current state of the PR to the above solution? Using feature gates

aagbotemi replied:
Alright, will do that.

nymius replied:
I think you are missing the `std` implementation using `rand` full features.

nymius replied:
> Looks good overall, not sure if any review comments are still unresolved.


@ValuedMammal I'm not sure we have fulfilled this requirements yet. Do you think we should roll a custom random range even if the user has `std` available?

aagbotemi commented:
> Sorry about the merge conflict.

Alright. I will rebase to resolve the conflict.

ValuedMammal reviewed (approved):
ACK ed9df3cbe183b3bc5cd7c5e7eb66873a6cd628a7

- You could make a note somewhere that this also includes an API breaking change in `CreatePsbtError`
- Following up on the no-std discussion, we should have default features disabled for `bdk_coin_select` dependency, but I ran into an issue trying to implement it - bitcoindevkit/coin-select#36

nymius reviewed (commented):
ACK ed9df3cbe183b3bc5cd7c5e7eb66873a6cd628a7

A little bit redundant now, but worth vouching for regardless.

To complement @ValuedMammal, I think is enough to change commit ed9df3cbe183b3bc5cd7c5e7eb66873a6cd628a7 message from `feat: enable random anti-fee sniping` to `feat!: enable random anti-fee sniping` and comment about the rationale for the variant additions on the body of the commit message.

aagbotemi commented:
> make a note somewhere that this also includes an API breaking change in CreatePsbtError

Noted. Thank you.

> Following up on the no-std discussion, we should have default features disabled for bdk_coin_select dependency, but I ran into an issue trying to implement it - https://github.com/bitcoindevkit/coin-select/issues/36

no-std support is blocked by bdk_coin_select not having a no-std feature. I will fix it upstream.

>change commit https://github.com/bitcoindevkit/bdk-tx/commit/ed9df3cbe183b3bc5cd7c5e7eb66873a6cd628a7 message from feat: enable random anti-fee sniping to feat!: enable random anti-fee sniping and comment about the rationale for the variant additions on the body of the commit message.

Thank you for the suggestion. I will fix it

nymius reviewed (approved):
ACK 179860e9bbe4c763296d275d47462a192ca2a903

ValuedMammal commented:
ACK 179860e9bbe4c763296d275d47462a192ca2a903

evanlinjin commented on src/selection.rs:
````diff
@@ -52,6 +57,57 @@ pub struct PsbtParams {
+    /// ```
+    ///
+    /// See [BIP326](https://github.com/bitcoin/bips/blob/master/bip-0326.mediawiki) for more details.
+    pub enable_anti_fee_sniping: bool,
````
This parameter quietly expects `fallback_locktime` to be a block height, otherwise we get `InvalidLockTime`.

We should design our structures so that a height is always provided when we enable AFS so that `InvalidLockTime` is no longer needed. For example:

```rust
anti_fee_sniping: Option<AfsParams>,

struct AfsParams { pub tip_height: absolute::Height }
```

evanlinjin reviewed (commented):
The AFS logic operates on an already-built transaction, so it shouldn't be directly tied into `create_psbt`.

The only reason we have `create_psbt_with_rng` is to support this unnecessary coupling. It's also the only creator of the `InvalidLockTime` and `UnsupportedVersion` variants of `CreatePsbtError`.

### Suggested Shape

* `apply_anti_fee_sniping` already exists in the crate root, we need to make it public!
* Have a convenience method on `Selection`:

```rust
impl Selection {
    pub fn apply_anti_fee_sniping(
        &self,
        psbt: &mut Psbt,
        tip_height: absolute::Height,
        rng: &mut impl RngCore,
    ) -> Result<(), AntiFeeSnipingError> {
        apply_anti_fee_sniping(&mut psbt.unsigned_tx, &self.inputs, tip_height, rng)
    }
}
```
